acra/acra-in-depth/data-structures/_index.md

```diff
@@ -2,3 +2,180 @@
+  `Encrypted_Random_Key = SMessage(RK, Throwaway_Private_Key, Acra_Public_Key)`.
+- Encrypts the payload with [Secure Cell]({{< ref "themis/crypto-theory/cryptosystems/secure-cell.md" >}}) in Seal mode:<br/>
+  `Encrypted_Data = SCell(RK, payload)`.
+- Erases/fills with zeros memory area of the `RK`.
+- Calculates the encrypted payload length and transforms it into little endian 8 bytes long (`Data_Length`).
+- Connects attributes together as described in the original formula.
+- Erases/fills with zeros the memory area containing the `Throwaway_Keypair` and original payload.
+
+We recommend you to check out AcraStruct [examples](https://github.com/cossacklabs/acra/tree/master/examples) and try the [⚙️Acra Engineering Demo⚙️](https://github.com/cossacklabs/acra-engineering-demo/#what-is-this).
+
+You can check your own AcraWriter implementation with our online [AcraStruct validator]({{< ref "acra/acra-in-depth/data-structures/acrastruct-validator.md" >}}) and try to decrypt generated AcraStructs.
+
+### Decryption
+
+AcraServer, upon receiving and detecting valid AcraStruct, is able to:
+
+- Extracts Throwaway Public Key (`TPK`).
+- Decrypts asymmetric envelope with `TPK` and Acra's Private Key (or [Zone key]({{< ref "acra/guides/integrating-acra-server-into-infrastructure#with-zones-INVALID">}})).
+- Extracts Random Key (`RK`) for Secure Cell container out of a decrypted envelope;
+- Decrypts Secure Cell, extracts payload;
+- Reconstructs database answer in such a way that AcraStruct is replaced by decrypted data.
+
+Check implementation in [decryptor/base/utils.go](https://github.com/cossacklabs/acra/blob/master/decryptor/base/utils.go#L29).
+
+
+## Storage Models
+
+There are two storage model modes used in Acra: WholeCell and InjectedCell.
+
+In WholeCell mode, CryptoEnvelope ([AcraStruct]({{< ref "acra/acra-in-depth/data-structures/#acrastruct" >}}) or [AcraBlock]({{< ref "acra/acra-in-depth/data-structures/#acrablock" >}})) represents a complete piece of data (i.e. database cell, a file, or some data transmitted into [AcraTranslator]({{< ref "acra/configuring-maintaining/general-configuration/acra-translator.md#-INVALID" >}})). In this mode it is expected that the encrypted data will look something like:
+
+1. `<CryptoEnvelope>`,
+2. `<CryptoEnvelope>`,
+3. `<CryptoEnvelope>`.
+
+In InjectedCell mode, CryptoEnvelope is stored inside some piece of data, i.e. inside some file or in a database cell with a file inside, with CryptoEnvelope as a piece of that file, not the whole file. In this mode, the encrypted data will look something like this:
+
+1. `<Some CryptoEnvelope data, some other CryptoEnvelope data>`,
+2. `<CryptoEnvelope>`,
+3. `<File containing CryptoEnvelope alongside other data>`.
```

Comment:
This is very confusing. Do all three parts contain CryptoEnvelope? Or just one? Is the order important? How is this different from three CryptoEnvelopes above? So many questions, so few answers...

The point you need to carry across is that in WholeCell mode encrypted data looks like this:

```
+-------------------------+
|                         |
| AcraStruct or AcraBlock |
|                         |
+-------------------------+
```

That is, exactly one AcraStruct which is the only and entire thing a data piece contains.

While InjectedCell may look like this:

```
+-------------------------+
|  Arbitrary prefix data  |
+-------------------------+
|                         |
| AcraStruct or AcraBlock |
|                         |
+-------------------------+
|      Some more data     |
| that Acra retains as is |
+-------------------------+
|    Another AcraStruct   |
+-------------------------+
|    More non-Acra data   |
+-------------------------+
```

Where Acra will find and decrypt all embedded AcraStructs while leaving everything else intact.

Some specific examples would also help, I believe. IDK, some Protobuf with three blobs where only one of them in actually encrypted.